app/assets/src/components/views/PipelineViz/PipelineViz.jsx

```diff
@@ -48,29 +50,95 @@ class PipelineViz extends React.Component {
       sidebarVisible: false,
       sidebarParams: {},
       hovered: false,
+      graphData: props.graphData,
     };
   }
 
   componentDidMount() {
+    const { updateInterval } = this.props;
     this.drawGraphs();
     window.addEventListener("resize", this.handleWindowResize);
+    // Only set update loop if the pipeline is still running.
+    if (!this.pipelineIsFinished()) {
+      this.updateLoop = setInterval(() => this.updateGraphs(), updateInterval);
+    }
   }
 
   componentWillUnmount() {
     window.removeEventListener("resize", this.handleWindowResize);
+    clearInterval(this.updateLoop);
+  }
+
+  async updateGraphs() {
+    const { sample, pipelineRun } = this.props;
+    const { graphData } = this.state;
+
+    const oldGraphData = graphData;
+    const newGraphData = await getGraph(
+      sample.id,
+      pipelineRun.version.pipeline
+    );
+    this.setState(
+      {
+        graphData: newGraphData,
+      },
+      () => this.pipelineIsFinished() && clearInterval(this.updateLoop)
+    );
+    const updates = diff(oldGraphData, newGraphData);
+    if (updates.stages) {
+      Object.keys(updates.stages).forEach(stageIndexStr => {
```

Comment:
Why is the stageIndex a str?

Reply:
The diff library I'm using stores changes in arrays as objects where the key is the string of the index https://www.npmjs.com/package/deep-object-diff

Reply:
blah... but ok